File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.java


import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		n = sc.nextInt();
		m = sc.nextInt();
		k = sc.nextInt(); // 회전 갯수

		grid = new int[n][m];
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				grid[i][j] = sc.nextInt();
			}
		}

		int kk = 0;

		arr = new int[k][4];
		sel = new int[k][4];

		while (kk < k) {
			int r = sc.nextInt();
			int c = sc.nextInt();
			int s = sc.nextInt();

			arr[kk][0] = r - s - 1;
			arr[kk][1] = c - s - 1;
			arr[kk][2] = r + s - 1;
			arr[kk][3] = c + s - 1;
			// 0 1 , 4 5 범위를 회전할 거임.

			kk++;
		}
		ans = Integer.MAX_VALUE;
		visited = new boolean[k];
		perm(0);
		System.out.println(ans);

	}

	static int k;
	static int[][] arr;
	static int[][] sel;
	static boolean[] visited;
	static boolean[][] check;
	static int n;
	static int m;
	static int[][] grid;
	static int[][] cgrid;
	static int[] range;
	static int ans;

	private static void perm(int idx) {

		if (idx == k) {
//			System.out.println(Arrays.deepToString(sel));

			cgrid = new int[n][m];

			for (int i = 0; i < n; i++) {
				for (int j = 0; j < m; j++) {
					cgrid[i][j] = grid[i][j];
				}
			}

			for (int i = 0; i < k; i++) {
				range = sel[i];
				check = new boolean[n][m];
				int r1 = range[0]; // 0
				int r2 = range[2]; // 4
				int c1 = range[1]; // 1
				int c2 = range[3]; // 5
				for (int r = r1; r <= r2; r++) {
					for (int c = c1; c <= c2; c++) {
						if (!check[r][c]) {
							rotation(r, c, r2, c2);
//							print(check);
//							print(cgrid);
							r2--;
							c2--;
						}
					}
				}

//				print(cgrid);
			}

			for (int i = 0; i < n; i++) {
				int sum = 0;
				for (int j = 0; j < m; j++) {
					sum += cgrid[i][j];
				}
				ans = Math.min(sum, ans);
			}

			return;
		}
		for (int i = 0; i < k; i++) {
			if (!visited[i]) {
				sel[idx] = arr[i].clone();
				visited[i] = true;
				perm(idx + 1);
				visited[i] = false;
			}
		}

	}

	private static void print(boolean[][] check2) {
		System.out.println("------------");
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				System.out.print(check2[i][j] + " ");
			}
			System.out.println();
		}

	}

	private static void print(int[][] cgrid2) {
		System.out.println("------------");
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				System.out.print(cgrid2[i][j] + " ");
			}
			System.out.println();
		}

	}

	private static void rotation(int r1, int c1, int r2, int c2) {
		if(r1==r2 && c1 == c2) {
			return;
		}
		int first = cgrid[r1][c1];
		for (int i = r1; i < r2; i++) {
			cgrid[i][c1] = cgrid[i + 1][c1];
			check[i][c1] = true;
		}

		for (int j = c1; j < c2; j++) {
			cgrid[r2][j] = cgrid[r2][j + 1];
			check[r2][j] = true;

		}
		for (int i = r2; i > r1; i--) {
			cgrid[i][c2] = cgrid[i - 1][c2];
			check[i][c2] = true;
		}
		for (int j = c2; j > c1; j--) {
			cgrid[r1][j] = cgrid[r1][j - 1];
			check[r1][j] = true;

		}

		cgrid[r1][c1 + 1] = first;

	}

}

File: 백준/Gold/17406. 배열 돌리기 4/배열 돌리기 4.py
'''
b17406

문제 설명
    시작점, 끝점, size로 배열을 회전하라
    한칸씩 시계방향으로 회전.
    아 근데 순열이 섞인...

'''

n,m,order_num = map(int, input().split())
grid = [list(map(int, input().split())) for i in range(n)]
grid_origin = [_[:] for _ in grid]
size_list = []
for i in range(order_num):
    r,c, s = map(int, input().split())
    size_list.append((r,c,s))

row = [0,1,0,-1]
col = [1,0,-1,0]

def rotation(r1,r2,c1,c2):
    d = 0
    grid_copy = [[0]*m for i in range(n)]
    or1, or2,oc1,oc2 = r1,r2,c1,c2

    while True:
        r = r1
        c = c1 # 시작점
        if(r == or1+(or2-or1)//2 and c == oc1+ (oc2-oc1)//2):
            grid_copy[r][c] = grid[r][c]
            break

        while True:

            if not(r1<=r+row[d]<=r2 and c1<=c+col[d]<=c2) or grid_copy[r+row[d]][c+col[d]]!=0:
                d = (d+1)%4
            nr = r+row[d]
            nc = c+col[d]

            grid_copy[nr][nc] = grid[r][c]
            r = nr
            c = nc

            if(r==r1 and c ==c1):
                break


        r1+=1
        c1+=1
        r2-=1
        c2-=1

    for i in range(or1, or2 + 1):
        for j in range(oc1, oc2 + 1):
            grid[i][j] = grid_copy[i][j]

mini = 100 * 50 + 1
sel = [0]*order_num

visited = [False]*order_num

def perm(idx):
    global mini, grid
    if(idx == order_num):

        for s in sel:
            r,c,s = size_list[s]

            rotation(r - s - 1, r + s - 1, c - s - 1, c + s - 1)

        for _ in grid:
            mini = min(mini, sum(_))
        grid = [_[:] for _ in grid_origin]
        return

    for i in range(order_num):
        if not visited[i]:
            visited[i]= True
            sel[idx] = i
            perm(idx+1)
            visited[i] = False

perm(0)
print(mini)